components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java

```diff
@@ -258,4 +291,234 @@ private boolean canAuthenticate(OAuthClientAuthenticator oAuthClientAuthenticato
+                    } else {
+                        log.error(basicAuthErrorMessage);
+                        throw new OAuthClientAuthnException(basicAuthErrorMessage, OAuth2ErrorCodes.INVALID_CLIENT);
+                    }
+                } else {
+                    log.error(basicAuthErrorMessage);
+                    throw new OAuthClientAuthnException(basicAuthErrorMessage, OAuth2ErrorCodes.INVALID_CLIENT);
+                }
+            } else {
+                throw new OAuthClientAuthnException("Unable to find client id in the request",
+                        OAuth2ErrorCodes.INVALID_CLIENT);
+            }
+        } catch (ParseException e) {
+            throw new OAuthClientAuthnException("Error occurred while parsing the signed assertion",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+    }
+
+    /**
+     * Validate whether the request follows MTLS client authentication.
+     *
+     * @param request               Http servlet request.
+     * @param contentParam          Map of request body params.
+     * @return Whether the request follows MTLS client authentication.
+     * @throws OAuthClientAuthnException
+     */
+    private boolean isMTLSAuthentication(HttpServletRequest request, Map<String, List> contentParam)
+            throws OAuthClientAuthnException {
+
+        String certificate = Optional.ofNullable(IdentityUtil.getProperty(OAuthConstants.MTLS_AUTH_HEADER))
+                .orElse("CONFIG_NOT_FOUND");
+
+        String oauthClientID =  request.getParameter(OAuth.OAUTH_CLIENT_ID);
+        if (StringUtils.isBlank(OAuth.OAUTH_CLIENT_ID)) {
+            oauthClientID = (String) contentParam.get(OAuth.OAUTH_CLIENT_ID).get(0);
+        }
+        try {
+            String oauthClientSecretValue = request.getParameter(OAuth.OAUTH_CLIENT_SECRET);
+            String oauthJWTAssertion = request.getParameter(OAuthConstants.OAUTH_JWT_ASSERTION);
+            String oauthJWTAssertionType = request.getParameter(OAuthConstants.OAUTH_JWT_ASSERTION_TYPE);
+            String authorizationHeaderValue = request.getHeader(OAuthConstants.AUTHORIZATION_HEADER);
+            String x509Certificate = request.getHeader(certificate);
+            if (StringUtils.isEmpty(x509Certificate)) {
+                x509Certificate = IdentityUtil.convertCertificateToPEM((Certificate) request
+                        .getAttribute(certificate));
+            }
+            return (StringUtils.isNotEmpty(oauthClientID) && StringUtils.isEmpty(oauthClientSecretValue) &&
+                    StringUtils.isEmpty(oauthJWTAssertion) && StringUtils.isEmpty(oauthJWTAssertionType) &&
+                    StringUtils.isEmpty(authorizationHeaderValue) && x509Certificate != null &&
+                    OAuth2Util.parseCertificate(x509Certificate) != null);
+        } catch (CertificateException e) {
+            throw new OAuthClientAuthnException("Error occurred while parsing the certificate",
```

Comment:
will it be better to log the error here?

Reply:
Fixed in 1227a3c7eda7627ef2cf6a63f8e54714f18af3a9